PhysicsTools/NanoAOD/plugins/HoverEValueMapProducer.cc

```diff
@@ -0,0 +1,162 @@
+  float getEtaForEA(const T*) const;
+  void doHoverEPho(edm::Event&) const;
+};
+
+//
+// constants, enums and typedefs
+//
+
+//
+// static data member definitions
+//
+
+template <typename T>
+float HoverEValueMapProducer<T>::getEtaForEA(const T* obj) const {
+  return obj->eta();
+}
+
+template <>
+float HoverEValueMapProducer<pat::Photon>::getEtaForEA(const pat::Photon* ph) const {
+  return ph->superCluster()->eta();
+}
+
+template <typename T>
+void HoverEValueMapProducer<T>::produce(edm::StreamID streamID,
+                                        edm::Event& iEvent,
+                                        const edm::EventSetup& iSetup) const {
+  if ((typeid(T) == typeid(pat::Photon))) {
+    doHoverEPho(iEvent);
+  }
+}
+
+template <typename T>
+void HoverEValueMapProducer<T>::doHoverEPho(edm::Event& iEvent) const {}
+
+template <>
+void HoverEValueMapProducer<pat::Photon>::doHoverEPho(edm::Event& iEvent) const {
+  edm::Handle<edm::View<pat::Photon>> src;
+  iEvent.getByToken(src_, src);
+  edm::Handle<double> rho;
+  iEvent.getByToken(rho_, rho);
+
+  unsigned int nInput = src->size();
+
+  std::vector<float> HoverEQuadratic;
+  HoverEQuadratic.reserve(nInput);
+
+  for (const auto& obj : *src) {
+    auto hOverE = obj.hcalOverEcal();
+
+    auto quadratic_ea_hOverE = quadratic_ea_hOverE_->getQuadraticEA(fabs(getEtaForEA(&obj)));
+    auto linear_ea_hOverE = quadratic_ea_hOverE_->getLinearEA(fabs(getEtaForEA(&obj)));
+
+    float scale = relative_ ? 1.0 / obj.pt() : 1;
+
+    HoverEQuadratic.push_back(
+        scale * (std::max(0.0, hOverE - (quadratic_ea_hOverE * (*rho) * (*rho) + linear_ea_hOverE * (*rho)))));
+  }
+
+  std::unique_ptr<edm::ValueMap<float>> HoverEQuadraticV(new edm::ValueMap<float>());
+  edm::ValueMap<float>::Filler fillerHoverEQuadratic(*HoverEQuadraticV);
+  fillerHoverEQuadratic.insert(src, HoverEQuadratic.begin(), HoverEQuadratic.end());
+  fillerHoverEQuadratic.fill();
+
+  iEvent.put(std::move(HoverEQuadraticV), "HoEForPhoEACorr");
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+template <typename T>
+void HoverEValueMapProducer<T>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("src")->setComment("input physics object collection");
+  desc.add<bool>("relative")->setComment("compute relative HoverE instead of absolute one");
+  if ((typeid(T) == typeid(pat::Photon))) {
+    desc.add<edm::FileInPath>("QuadraticEAFile_HoverE")
+        ->setComment("txt file containing quadratic effective areas to be used for H/E pileup subtraction for photons");
+
+    desc.add<edm::InputTag>("rho")->setComment(
+        "rho to be used for effective-area based H/E pileup subtraction for photons");
+  }
+
+  std::string modname;
+  if (typeid(T) == typeid(pat::Photon))
+    modname += "Pho";
+  modname += "HoverEValueMapProducer";
+  descriptions.add(modname, desc);
```

Comment:
I think you can just do `descriptions.addWithDefaultLabel(desc);` and it will use the name of the template typedef.

Reply:
Ok, thanks for the suggestion.